extensions/amp-form/0.1/amp-form.js

```diff
@@ -209,39 +221,70 @@ export class AmpForm {
       this.setState_(FormState_.SUBMITTING);
       const isHeadOrGet = this.method_ == 'GET' || this.method_ == 'HEAD';
       let xhrUrl = this.xhrAction_;
-      if (isHeadOrGet) {
-        xhrUrl = addParamsToUrl(this.xhrAction_, this.getFormAsObject_());
+
+      const varSubPromises = [];
+      for (let i = 0; i < varSubsFields.length; i++) {
+        const variable = varSubsFields[i].getAttribute('default-value');
+        varSubPromises.push(
+            this.urlReplacement_.expandAsync(variable).then(value => {
+              varSubsFields[i].value = value;
+            }));
       }
-      this.xhr_.fetchJson(xhrUrl, {
-        body: isHeadOrGet ? undefined : new FormData(this.form_),
-        method: this.method_,
-        credentials: 'include',
-        requireAmpResponseSourceOrigin: true,
-      }).then(response => {
-        this.actions_.trigger(this.form_, 'submit-success', null);
-        // TODO(mkhatib, #6032): Update docs to reflect analytics events.
-        this.analyticsEvent_('amp-form-submit-success');
-        this.setState_(FormState_.SUBMIT_SUCCESS);
-        this.renderTemplate_(response || {});
-      }).catch(error => {
-        this.actions_.trigger(this.form_, 'submit-error', null);
-        this.analyticsEvent_('amp-form-submit-error');
-        this.setState_(FormState_.SUBMIT_ERROR);
-        this.renderTemplate_(error.responseJson || {});
-        rethrowAsync('Form submission failed:', error);
+      // Wait until all variables have been substituted or 100ms timeout.
+      return this.waitOnPromisesOrTimeout_(varSubPromises, 100).then(() => {
+        if (isHeadOrGet) {
+          xhrUrl = addParamsToUrl(this.xhrAction_, this.getFormAsObject_());
```

Comment:
Nit: This would be more readable if `xhrUrl` was declared closer to its usage and the `isHeadOrGet` conditional wasn't duped below for `body`. E.g.

```
let xhrUrl, body;
if (isHeadOrGet) {
  xhrUrl = addParamsToUrl(this.xhrAction_, this.getFormAsObject_());
  body = new FormData(this.form_);
} else {
  xhrUrl = this.xhrAction_;
}
return this.xhr_.fetchJson(xhrUrl, {
  body: body,
```